python/visualization.py

```diff
@@ -0,0 +1,1076 @@
+        else:
+            vol = get_boundary_volumes(boundary.thickness,boundary.direction,boundary.side)
+            ax = plot_volume(sim,ax,vol,output_plane,plotting_parameters=boundary_parameters)
+    return ax
+
+def plot_sources(sim,ax,output_plane=None,labels=False,source_parameters=None):
+    from meep.simulation import Volume
+
+    # consolidate plotting parameters
+    source_parameters = default_source_parameters if source_parameters is None else dict(default_source_parameters, **source_parameters)
+
+    label = 'source' if labels else None
+
+    for src in sim.sources:
+        vol = Volume(center=src.center,size=src.size)
+        ax = plot_volume(sim,ax,vol,output_plane,plotting_parameters=source_parameters,label=label)
+    return ax
+
+def plot_monitors(sim,ax,output_plane=None,labels=False,monitor_parameters=None):
+    from meep.simulation import Volume
+
+     # consolidate plotting parameters
+    monitor_parameters = default_monitor_parameters if monitor_parameters is None else dict(default_monitor_parameters, **monitor_parameters)
+
+    label = 'monitor' if labels else None
+
+    for mon in sim.dft_objects:
+        for reg in mon.regions:
+            vol = Volume(center=reg.center,size=reg.size)
+            ax = plot_volume(sim,ax,vol,output_plane,plotting_parameters=monitor_parameters,label=label)
+    return ax
+
+def plot_fields(sim,ax=None,fields=None,output_plane=None,field_parameters=None):
+
+    if fields is None:
+        return ax
+    
+    # user specifies a field component
+    if fields in [mp.Ex, mp.Ey, mp.Ez, mp.Hx, mp.Hy, mp.Hz]:
+        # Get domain measurements
+        if output_plane:
+            sim_center, sim_size = (output_plane.center, output_plane.size)
+        elif sim.output_volume:
+            sim_center, sim_size = mp.get_center_and_size(sim.output_volume)
+        else:
+            sim_center, sim_size = (sim.geometry_center, sim.cell_size)
+        
+        xmin = sim_center.x - sim_size.x/2
+        xmax = sim_center.x + sim_size.x/2
+        ymin = sim_center.y - sim_size.y/2
+        ymax = sim_center.y + sim_size.y/2
+        zmin = sim_center.z - sim_size.z/2
+        zmax = sim_center.z + sim_size.z/2
+
+        center = Vector3(sim_center.x,sim_center.y,sim_center.z)
+        cell_size = Vector3(sim_size.x,sim_size.y,sim_size.z)
+
+        if sim_size.x == 0:
+            # Plot y on x axis, z on y axis (YZ plane)
+            extent = [ymin,ymax,zmin,zmax]
+            xlabel = 'Y'
+            ylabel = 'Z'
+        elif sim_size.y == 0:
+            # Plot x on x axis, z on y axis (XZ plane)
+            extent = [xmin,xmax,zmin,zmax]
+            xlabel = 'X'
+            ylabel = 'Z'
+        elif sim_size.z == 0:
+            # Plot x on x axis, y on y axis (XY plane)
+            extent = [xmin,xmax,ymin,ymax]
+            xlabel = 'X'
+            ylabel = 'Y'
+        fields = sim.get_array(center=center, size=cell_size, component=fields)
```

Comment:
`get_array` must be called on all procs.

Reply:
That's the issue. I'll go through and fix all of those. Thanks so much!

Reply:
Is that requirement documented anywhere? We may want to add a page about using `mp.am_master()` with `matplotlib` and when `am_master()` shouldn't be used (i.e. with `get_array()`